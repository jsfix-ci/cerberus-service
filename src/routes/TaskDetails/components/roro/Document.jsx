import React from 'react';
import classNames from 'classnames';

// Utils
import { DocumentUtil, PersonUtil } from '../../../../utils';

import renderBlock from '../../helper/common';

const Account = ({ version, classModifiers }) => {
  const person = PersonUtil.get(version);
  const document = DocumentUtil.get(person);
  return (
    <div className={classNames('task-details-container', 'govuk-!-margin-bottom-2', classModifiers)}>
      <h3 className="govuk-heading-m govuk-!-margin-top-0">Document</h3>
      <div className="govuk-task-details-grid-column">
        {renderBlock('Travel document type', [DocumentUtil.docType(document)])}
        {renderBlock('Travel document number', [DocumentUtil.docNumber(document)])}
        {renderBlock('Travel document expiry', [DocumentUtil.docExpiry(document, true)])}
      </div>
    </div>
  );
};

export default Account;
