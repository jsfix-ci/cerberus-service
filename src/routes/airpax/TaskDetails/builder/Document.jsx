import React from 'react';
// Utils
import { getDate } from '../../utils/datetimeUtil';
import { getJourney, getArrivalTime } from '../../utils/movementUtil';

import {
  DocumentUtil,
  PersonUtil,
} from '../../utils';

import renderBlock from './helper/common';

const Document = ({ version }) => {
  const person = PersonUtil.get(version);
  const document = DocumentUtil.get(version.movement.person);
  const journey = getJourney(version);
  const validFromExpiry = document ? DocumentUtil.calculateExpiry(document.validFrom, getDate()) : 'Unknown';
  const validToExpiry = document ? DocumentUtil.calculateExpiry(document.expiry, getArrivalTime(journey)) : 'Unknown';

  return (
    <div className="task-details-container">
      <h3 className="govuk-heading-m govuk-!-margin-top-0">Document</h3>
      <div className="govuk-task-details-grid-column">
        {renderBlock('Type', [DocumentUtil.docType(document)])}
        {renderBlock('Number', [DocumentUtil.docNumber(document)])}
        {renderBlock('Document nationality', [DocumentUtil.docNationality(document, true)])}
        {renderBlock('Country of issue', [DocumentUtil.docCountry(document, true)])}
        {renderBlock('Valid from', [DocumentUtil.docValidity(document, true), validFromExpiry])}
        {renderBlock('Valid To', [DocumentUtil.docExpiry(document, true), validToExpiry])}
        {renderBlock('Name', [DocumentUtil.docName(person)])}
        {renderBlock('Date of birth', [PersonUtil.dob(person)])}
      </div>
    </div>
  );
};

export default Document;
