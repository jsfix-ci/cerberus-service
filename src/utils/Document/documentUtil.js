import moment from 'moment';
import lookup from 'country-code-lookup';
import {
  A_TITLE_CASE_TEXT,
  AN_TITLE_CASE_TEXT,
  AGO_TEXT,
  A_SMALL_TEXT,
  AN_SMALL_TEXT,
  AFTER_TRAVEL_TEXT,
  BEFORE_TRAVEL_TEXT,
  SHORT_DATE_FORMAT_ALT,
  UNKNOWN_TEXT,
  UTC_DATE_FORMAT,
} from '../constants';
import { formatField } from '../FieldFormat/fieldFormatterUtil';
import Common from '../Common/common';
import { getFormattedDate, validateDate } from '../Datetime/datetimeUtil';

const calculateExpiry = (passportExpiry, arrivalTime) => {
  if (!validateDate(passportExpiry) || !validateDate(arrivalTime)) {
    return UNKNOWN_TEXT;
  }
  const expiry = `${arrivalTime},${moment(passportExpiry).format(UTC_DATE_FORMAT)}`;
  return formatField('BOOKING_DATETIME', expiry)
    .split(', ')[1]
    .replace(BEFORE_TRAVEL_TEXT, AFTER_TRAVEL_TEXT)
    .replace(AGO_TEXT, BEFORE_TRAVEL_TEXT)
    .replace(A_SMALL_TEXT, A_TITLE_CASE_TEXT)
    .replace(AN_SMALL_TEXT, AN_TITLE_CASE_TEXT);
};

const hasDocument = (person) => {
  return !!person?.document;
};

const getDocument = (person) => {
  if (hasDocument(person)) {
    return person.document;
  }
  return null;
};

const getDocumentExpiryDate = (document) => {
  return document?.expiry || undefined;
};

const getDocumentExpiry = (document, taskDetails = false) => {
  const expiryPrefix = 'Expires';
  if (!document?.expiry) {
    return taskDetails ? UNKNOWN_TEXT : `${expiryPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}` : `${expiryPrefix} ${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentValidityDate = (document) => {
  return document?.validFrom || undefined;
};

const getDocumentValidity = (document, taskDetails = false) => {
  const validityPrefix = 'Valid from';
  if (!document?.validFrom) {
    return taskDetails ? UNKNOWN_TEXT : `${validityPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}` : `${validityPrefix} ${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentType = (document) => {
  if (!document?.type) {
    return UNKNOWN_TEXT;
  }
  return document.type;
};

const getDocumentNumber = (document) => {
  if (!document?.number) {
    return UNKNOWN_TEXT;
  }
  return document.number;
};

const getDocumentName = (person) => {
  if (!person?.name) {
    return UNKNOWN_TEXT;
  }
  const firstNames = person?.name?.first || UNKNOWN_TEXT;
  const lastName = person?.name?.last || UNKNOWN_TEXT;
  return `${lastName.toUpperCase()}, ${firstNames}`;
};

// TODO finish implementation once data flows through
const getDocumentIdentification = (document) => {
  if (!document) {
    return UNKNOWN_TEXT;
  }
  return UNKNOWN_TEXT;
};

const getDocumentCountryOfIssue = (document, taskDetails = false) => {
  const countryOfIssuePrefix = 'Issued by';
  if (!document?.countryOfIssue) {
    return taskDetails ? UNKNOWN_TEXT : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }
  if (taskDetails) {
    return lookup.byIso(document.countryOfIssue) !== null
      ? `${lookup.byIso(document.countryOfIssue).country} (${Common.iso3Code(document.countryOfIssue)})`
      : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }

  return lookup.byIso(document.countryOfIssue) !== null
    ? `${countryOfIssuePrefix} ${Common.iso3Code(document.countryOfIssue)}`
    : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
};

const getDocumentCountryOfIssueCode = (document) => {
  if (!document?.countryOfIssue) {
    return UNKNOWN_TEXT;
  }
  return Common.iso3Code(document.countryOfIssue);
};

const getDocumentNationality = (document, taskDetails = false) => {
  return getDocumentCountryOfIssue(document, taskDetails);
};

const getDocumentDOB = (document) => {
  if (!document?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(document?.dateOfBirth, SHORT_DATE_FORMAT_ALT);
};

const DocumentUtil = {
  get: getDocument,
  docExpiry: getDocumentExpiry,
  docExpiryDate: getDocumentExpiryDate,
  docValidity: getDocumentValidity,
  docValidityDate: getDocumentValidityDate,
  docType: getDocumentType,
  docNumber: getDocumentNumber,
  docName: getDocumentName,
  docIdentification: getDocumentIdentification,
  docCountry: getDocumentCountryOfIssue,
  docCountryCode: getDocumentCountryOfIssueCode,
  docNationality: getDocumentNationality,
  docDOB: getDocumentDOB,
  calculateExpiry,
};

export default DocumentUtil;

export {
  getDocument,
  getDocumentExpiry,
  getDocumentExpiryDate,
  getDocumentValidity,
  getDocumentValidityDate,
  getDocumentType,
  getDocumentNumber,
  getDocumentName,
  getDocumentIdentification,
  getDocumentCountryOfIssue,
  getDocumentCountryOfIssueCode,
  getDocumentNationality,
  getDocumentDOB,
  calculateExpiry,
};
