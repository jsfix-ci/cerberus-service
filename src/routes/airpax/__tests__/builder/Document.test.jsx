import React from 'react';
import renderer from 'react-test-renderer';

import Document from '../../TaskDetails/builder/Document';
import taskDetailsData from '../../__fixtures__/taskData_AirPax_TaskDetails.fixture.json';

describe('Document', () => {
  it('should render the document component', () => {
    const tree = renderer.create(<Document version={taskDetailsData.versions[1]} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
