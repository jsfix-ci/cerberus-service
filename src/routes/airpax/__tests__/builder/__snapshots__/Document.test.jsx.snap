// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document should render the document component 1`] = `
<div
  className="task-details-container"
>
  <h3
    className="govuk-heading-m govuk-!-margin-top-0"
  >
    Document
  </h3>
  <div
    className="govuk-task-details-grid-column"
  >
    <div
      className="govuk-!-margin-bottom-2"
    >
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Type
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__bold"
      >
        Unknown
      </p>
    </div>
    <div
      className="govuk-!-margin-bottom-2"
    >
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Number
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__bold"
      >
        Unknown
      </p>
    </div>
    <div
      className="govuk-!-margin-bottom-2"
    >
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Document nationality
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__bold"
      >
        Unknown
      </p>
    </div>
    <div
      className="govuk-!-margin-bottom-2"
    >
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Country of issue
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__bold"
      >
        Unknown
      </p>
    </div>
    <div
      className="govuk-!-margin-bottom-2"
    >
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Valid from
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__bold"
      >
        Unknown
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Unknown
      </p>
    </div>
    <div
      className="govuk-!-margin-bottom-2"
    >
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Valid To
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__bold"
      >
        Unknown
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Unknown
      </p>
    </div>
    <div
      className="govuk-!-margin-bottom-2"
    >
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Name
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__bold"
      >
        MESTA, GEMMA
      </p>
    </div>
    <div
      className="govuk-!-margin-bottom-2"
    >
      <p
        className="govuk-!-margin-bottom-0 font__light"
      >
        Date of birth
      </p>
      <p
        className="govuk-!-margin-bottom-0 font__bold"
      >
        Unknown
      </p>
    </div>
  </div>
</div>
`;
